river/expert/bandit.py

```diff
@@ -0,0 +1,269 @@
+
+        # Check that the model and the metric are in accordance
+        for model in models:
+            if not metric.works_with(model):
+                raise ValueError(f"{metric.__class__.__name__} metric can't be used to evaluate a " +
+                                 f'{model.__class__.__name__}')
+        super().__init__(models)
+        self.reward_scaler = copy.deepcopy(reward_scaler)
+        self.metric = copy.deepcopy(metric)
+        self.print_every = print_every
+
+        self.save_metric_values = save_metric_values
+        if save_metric_values:
+            self.metric_values: typing.List = []
+
+        self.save_percentage_pulled = save_percentage_pulled
+        if save_percentage_pulled:
+            self.store_percentage_pulled: typing.List = []
+
+        # Initializing bandits internals
+        self._n_arms = len(models)
+        self._n_iter = 0 # number of times learn_one is called
+        self._N = np.zeros(self._n_arms, dtype=np.int)
+        self._average_reward = np.zeros(self._n_arms, dtype=np.float)
+
+    @abc.abstractmethod
+    def _pull_arm(self):
+        pass
+
+    @abc.abstractmethod
+    def _update_arm(self, arm, reward):
+        pass
+
+    @abc.abstractmethod
+    def _pred_func(self, model):
+        pass
+
+    @property
+    def _best_model_idx(self):
+        # average reward instead of cumulated (otherwise favors arms which are pulled often)
+        return np.argmax(self._average_reward)
+
+    @property
+    def best_model(self):
+        return self[self._best_model_idx]
+
+    @property
+    def percentage_pulled(self):
+        percentages = self._N / sum(self._N)
+        return percentages
+
+    def predict_one(self, x):
+        best_arm = self._pull_arm()
+        y_pred = self._pred_func(self[best_arm])(x)
+        return y_pred
+
+    def learn_one(self, x, y):
+        chosen_arm = self._pull_arm()
+        chosen_model = self[chosen_arm]
+
+        y_pred = chosen_model.predict_one(x)
+        self.metric.update(y_pred=y_pred, y_true=y)
+        chosen_model.learn_one(x=x, y=y)
+
+        # Update bandit internals (common to all bandit)
+        reward = self._compute_scaled_reward(y_pred=y_pred, y_true=y)
+        self._n_iter += 1
+        self._N[chosen_arm] += 1
+        self._average_reward[chosen_arm] += (1.0 / self._N[chosen_arm]) * \
+                                            (reward - self._average_reward[chosen_arm])
+
+        # Specific update of the arm for certain bandit class
+        self._update_arm(chosen_arm, reward)
+
+        if self.print_every:
+            if (self._n_iter % self.print_every) == 0:
+                self._print_info()
+
+        if self.save_percentage_pulled:
+            self.store_percentage_pulled += [self.percentage_pulled]
+
+        if self.save_metric_values:
+            self.metric_values += [self.metric._eval(y_pred, y)]
+
+        return self
```

Comment:
I think that a good design would be to have a `_learn_one` method with returns `self.percentage_pulled` and `self.metric._eval(y_pred, y)`. That way, power users can use `_learn_one` to collect information and display/analyse it in some way. I don't think we should be storing a trace of the values.

Reply:
Ok, tell me if the way I've done it suits you

Reply:
Yes it looks good. But now that I think about it, you don't need to return `percentage_pulled` as it's already a property. Maybe you can return the pulled model? I guess the `_learn_one` method will change in the future depending on downstream usage. Also can just remove the `self` in the return.